### PR TITLE
Add engine_kwargs support to dask.dataframe.to_sql

### LINKED ISSUE #8596: Add `engine_kwargs` argument to `dask.dataframe.to_sql`
This question came up on the [Dask Discourse](https://dask.discourse.group/t/how-to-provide-ssl-details-in-the-ddf-to-sql-function/206). We were able to come up with a solution, but noticed the [`dask.dataframe.read_sql_table`](https://docs.dask.org/en/latest/generated/dask.dataframe.read_sql_table.html) accepts an `engine_kwargs` argument while [`dask.dataframe.to_sql`](https://docs.dask.org/en/latest/generated/dask.dataframe.to_sql.html) does not. Adding an `engine_kwargs` argument would be a nice feature that would make using `to_sql` less confusing.

### LINKED ISSUE #8622: Broken usage of capsys for inspecting logging output
**What happened**:

1. We have one unit test that uses capsys to make assertions about emitted logs: [`dask/dataframe/io/tests/test_sql.py`](https://github.com/dask/dask/blob/2022.01.0/dask/dataframe/io/tests/test_sql.py#L410-L425)
1. Add a second unit test that does the same
1. You can't have more than one test using that pattern, the first one works and anything after does not
    - add new test after old test: old test passes, new test fails
    - skip old test: new test passes
    - add new test before old test: new test passes, old test fails

**What you expected to happen**:

We can have as many tests as we want making assertions about logs.

**Minimal Complete Verifiable Example**:

```python
# repro.py
import sqlalchemy as sa
from tempfile import NamedTemporaryFile

def do_something():
    with NamedTemporaryFile() as f:
        engine = sa.create_engine(f'sqlite:///{f.name}', echo=True)
        engine.execute("CREATE TABLE t (id INTEGER);")

def test1(capsys):
    do_something()
    out, err = capsys.readouterr()
    assert 'CREATE' in out

def test2(capsys):
    do_something()
    out, err = capsys.readouterr()
    assert 'CREATE' in out
```

Usage:
```sh
$ pytest -v repro.py -k test1 # passes
$ pytest -v repro.py -k test2 # passes
$ pytest -v repro.py # test1 passes, test2 fails
```

<details> <summary> inspected stderr of failed test case </summary>
<pre>
Traceback (most recent call last):
  File "/usr/lib/python3.8/logging/__init__.py", line 1088, in emit
    stream.write(msg + self.terminator)
ValueError: I/O operation on closed file.
Call stack:
  File "/dask/venv/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  File "/dask/venv/lib/python3.8/site-packages/_pytest/config/__init__.py", line 185, in console_main
    code = main()
  File "/dask/venv/lib/python3.8/site-packages/_pytest/config/__init__.py", line 162, in main
    ret: Union[ExitCode, int] = config.hook.pytest_cmdline_main(
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/dask/venv/lib/python3.8/site-packages/_pytest/main.py", line 316, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/dask/venv/lib/python3.8/site-packages/_pytest/main.py", line 269, in wrap_session
    session.exitstatus = doit(config, session) or 0
  File "/dask/venv/lib/python3.8/site-packages/_pytest/main.py", line 323, in _main
    config.hook.pytest_runtestloop(session=session)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/dask/venv/lib/python3.8/site-packages/_pytest/main.py", line 348, in pytest_runtestloop
    item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/dask/venv/lib/python3.8/site-packages/_pytest/runner.py", line 109, in pytest_runtest_protocol
    runtestprotocol(item, nextitem=nextitem)
  File "/dask/venv/lib/python3.8/site-packages/_pytest/runner.py", line 126, in runtestprotocol
    reports.append(call_and_report(item, "call", log))
  File "/dask/venv/lib/python3.8/site-packages/_pytest/runner.py", line 215, in call_and_report
    call = call_runtest_hook(item, when, **kwds)
  File "/dask/venv/lib/python3.8/site-packages/_pytest/runner.py", line 254, in call_runtest_hook
    return CallInfo.from_call(
  File "/dask/venv/lib/python3.8/site-packages/_pytest/runner.py", line 311, in from_call
    result: Optional[TResult] = func()
  File "/dask/venv/lib/python3.8/site-packages/_pytest/runner.py", line 255, in <lambda>
    lambda: ihook(item=item, **kwds), when=when, reraise=reraise
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/dask/venv/lib/python3.8/site-packages/_pytest/runner.py", line 162, in pytest_runtest_call
    item.runtest()
  File "/dask/venv/lib/python3.8/site-packages/_pytest/python.py", line 1641, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/dask/venv/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/dask/venv/lib/python3.8/site-packages/_pytest/python.py", line 183, in pytest_pyfunc_call
    result = testfunction(**testargs)
  File "/dask/repro.py", line 23, in test2
    do_something()
  File "/dask/repro.py", line 13, in do_something
    engine.execute("CREATE TABLE t (id INTEGER);")
  File "<string>", line 2, in execute
  File "/dask/venv/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 401, in warned
    return fn(*args, **kwargs)
  File "/dask/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3146, in execute
    return connection.execute(statement, *multiparams, **params)
  File "/dask/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1274, in execute
    return self._exec_driver_sql(
  File "/dask/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1578, in _exec_driver_sql
    ret = self._execute_context(
  File "/dask/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1755, in _execute_context
    self._log_info(statement)
  File "/dask/venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 136, in _log_info
    self.engine.logger.info(message, *arg, **kw)
  File "/dask/venv/lib/python3.8/site-packages/sqlalchemy/log.py", line 117, in info
    self.log(logging.INFO, msg, *args, **kwargs)
  File "/dask/venv/lib/python3.8/site-packages/sqlalchemy/log.py", line 163, in log
    self.logger._log(level, msg, args, **kwargs)
Message: 'CREATE TABLE t (id INTEGER);'
</pre>
</details>

**Anything else we need to know?**: the issue manifested in #8609.

**Environment**:

- SQLAlchemy verison: 1.4.31
- pytest version: 6.2.5
- Dask version: main
- Python version: 3.8.10
- Operating System: Ubuntu 20.04
- Install method (conda, pip, source): source

### PR DESCRIPTION
**Pull request by amirkdv**
- [x] Closes #8596 
- [x] Closes #8622
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### What's tested

A new test case covers `to_sql(..., engine_kwargs=...)` using `echo=True` as the engine argument.

### Use case: SSL

This was the original [use case] that motivated #8596: to force SSL on the db connection. Whether the new `engine_kwargs` argument is helpful/necessary for this use case depends on the db driver used by SQLAlchemy:

* [MySQL] (helpful): either use `connect_args={'ssl': ...}` as engine kwarg or add `?ssl_cert=...&ssl_key=...` to URI.
* [psycopg2] (not helpful): must use `?sslmode=require` in connection URI, not supported as engine argument.
* [pg8000] (necessary): must use `connect_args={'ssl_context': ...}` as engine kwarg.

[use case]: https://github.com/coiled/dask-community/issues/186
[MySQL]: https://docs.sqlalchemy.org/en/14/dialects/mysql.html#ssl-connections
[psycopg2]: https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#ssl-connections
[pg8000]: https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#pg8000-ssl

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by jsignell** [comment]
ok to test

**Comment by ncclementi** [comment]
@amirkdv checking in here. It looks like we are having legit CI failures for ubuntu and macos (see https://github.com/dask/dask/runs/4909206666?check_suite_focus=true#step:6:21240), and for some reason we are having bunch of problems on windows all related to parquet.

**Comment by amirkdv** [comment]
@ncclementi Thanks for the ping. I had successfully run my new test locally but didn't dig through CI logs to see it failing when run in the full suite. Sorry about that!

There are a bunch of tests failing:
- some are failing [in main as well](https://github.com/dask/dask/runs/4940478147?check_suite_focus=true): windows 3.7, 3.9 and ubuntu 3.9 and macos 3.9
- all the other ones have to do with a new issue I discovered: #8622  **TL;DR**: our usage of capsys for making assertions about logs (which I took from an existing test case) is not right. But it fails only fails once you have >1 test case that does that!

I've pushed a fix for the new issue to this branch. Let me know if you want me to separate it into a separate PR.

File: dask/dataframe/io/tests/test_sql.py [hunk 1]
```diff
@@ -519,3 +519,11 @@ def test_to_sql_kwargs():
             TypeError, match="to_sql\\(\\) got an unexpected keyword argument 'unknown'"
         ):
             ddf.to_sql("test", uri, unknown=None)
+
+
+def test_to_sql_engine_kwargs(caplog):
```
Comment (scharlottej13) [inline dask/dataframe/io/tests/test_sql.py:524]: I'm wondering if this test should be more extensive like `test_extra_connection_engine_keywords` and also do a `read_sql_table` to ensure the result is the same.

**Review by scharlottej13 — CHANGES_REQUESTED** [review]
Thank you @amirkdv for your work on this! It's looking good, I have a couple suggestions for the test you wrote.

**Comment by scharlottej13** [comment]
>**TL;DR**: our usage of capsys for making assertions about logs (which I took from an existing test case) is not right. But it only fails once you have >1 test case that does that!
> 
> I've pushed a fix for the new issue to this branch. Let me know if you want me to separate it into a separate PR.

Thank you for adding these changes as well!

File: dask/dataframe/io/tests/test_sql.py (hunk 1 above)
Comment (amirkdv) [reply dask/dataframe/io/tests/test_sql.py:557]: Thanks! I added a few more scenarios that could be affected by the new code but weren't covered before, namely `echo=False` and `assert read(write(X)) == X`.

**Comment by amirkdv** [comment]
Thanks for the review  @scharlottej13! I addressed your comments, and rebased to resolve merge conflicts with main.

**Comment by amirkdv** [comment]
Not sure what's happening in CI; there's all sorts of new failures I hadn't seen before. But I've confirmed that `test_to_sql_engine_kwargs` is passing in all 9 environments.

**Comment by scharlottej13** [comment]
> Not sure what's happening in CI; there's all sorts of new failures I hadn't seen before. But I've confirmed that `test_to_sql_engine_kwargs` is passing in all 9 environments.

I'm not quite sure either, there is a failure on ubuntu-3.9 in `distributed/utils_test.py` and on macos 3.9 in `fastparquet`. @ncclementi do you have any ideas on how we can investigate these?

**Comment by ncclementi** [comment]
I think these might still be related to the sqlalchemy issue https://github.com/dask/dask/issues/7406 but not quite sure since that one seemed to be resolved 2 days ago, and @amirkdv merged main less dan a day ago. 

@jsignell Do you know what could be happening here?

**Comment by amirkdv** [comment]
For convenience:
- here are [CI tests](https://github.com/dask/dask/runs/4957175605?check_suite_focus=true) for my branch off commit in main: 7edfe245 (failing in similar ways + timeouts/cancels).
- current main branch has only one extra commit 193998cef which seems to be doing better: [CI tests](https://github.com/dask/dask/runs/4967669517?check_suite_focus=true) (in progress).

**Comment by jsignell** [comment]
Ah sorry I had missed the ping over here. There were some CI failures last week while we dealt with the latest pandas release. Things should be in better shape now. I'm going to re-trigger the tests now to make sure that we are all good.

**Comment by jsignell** [comment]
This is in @amirkdv! Thanks!

**Comment by ian-r-rose** [comment]
Thanks @amirkdv, and @jsignell and @scharlottej13 for reviewing!
